app/main/views/frameworks.py

```diff
@@ -329,31 +330,36 @@ def framework_supplier_declaration(framework_slug, section_id=None):
                     current_user.email_address
                 )
 
-                next_section = content.get_next_editable_section_id(section_id)
-                if next_section:
-                    return redirect(
-                        url_for('.framework_supplier_declaration',
-                                framework_slug=framework['slug'],
-                                section_id=next_section))
+                if next_section and not request.form.get('save_and_return_to_overview', False):
+                    # Go to the next section.
+                    return redirect(url_for(
+                        '.framework_supplier_declaration',
+                        framework_slug=framework['slug'],
+                        section_id=next_section.id
+                    ))
                 else:
-                    url = "{}/declaration_complete".format(
-                        url_for('.framework_dashboard',
-                                framework_slug=framework['slug']))
-                    flash(url, 'declaration_complete')
-                    return redirect(
-                        url_for('.framework_dashboard',
-                                framework_slug=framework['slug']))
+                    # Otherwise they may have completed their declaration. Check.
+                    if not next_section:
+                        # They have. Write the flash message.
+                        url = "{}/declaration_complete".format(
+                            url_for('.framework_dashboard', framework_slug=framework['slug'])
+                        )
+                        flash(url, 'declaration_complete')
```

Comment:
I don't think we're doing this here - this flash message is to accompany the final "Make declaration" button-press, done from the overview.

I think you can safely obliterate all machinery for setting flash messages from this view.